### DISCUSSION TITLE
Semaphore behind nginx proxy

### DISCUSSION THREAD

**Discussion by tapasmallickin** [semaphoreui/semaphore#2192] [Q&A]
We have semaphore (v2.10.11) deployed as a docker container with following parameters:
ports: 
  - 4000:3000
environment:
  SEMAPHORE_WEB_ROOT: /semaphore

And it is working fine.

We have deployed nginx and configured (on port 8080) as a reverse proxy with following:
location /semaphore/ {
                proxy_pass http://127.0.0.1:4000/;
                proxy_set_header Host $http_host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Proto $scheme;
                proxy_buffering off;
                proxy_request_buffering off;
                rewrite ^/semaphore/(.*) /$1 break;
        }

But we are getting an empty page with following error in nginx log:
127.0.0.1 - - [15/Jul/2024:08:51:36 +0000] "GET /semaphore/js/app.3c43688b.js HTTP/1.1" 404 10 "http://localhost:8080/semaphore/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36" "-"
127.0.0.1 - - [15/Jul/2024:08:51:36 +0000] "GET /semaphore/js/chunk-vendors.46072356.js HTTP/1.1" 404 10 "http://localhost:8080/semaphore/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36" "-"
127.0.0.1 - - [15/Jul/2024:08:51:36 +0000] "GET /semaphore/css/app.bec6951a.css HTTP/1.1" 404 10 "http://localhost:8080/semaphore/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36" "-"
127.0.0.1 - - [15/Jul/2024:08:51:36 +0000] "GET /semaphore/css/chunk-vendors.a54f31b2.css HTTP/1.1" 404 10 "http://localhost:8080/semaphore/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36" "-"

Looking forward for any assistance here.

**Comment by tapasmallickin** [comment]
I believe it is a standard configuration and someone already implemented such. Your assistance is appreciated.

**Comment by JtheBAB** [comment]
This is my config:

```
server {
        listen 80;
        return 301 https://$host$request_uri;
}

server {
        listen 443 ssl http2;
        server_namefqdn;

        access_log  /var/log/nginx/hostname_access.log;
        error_log  /var/log/nginx/hostname_error.log;


        ssl_certificate      /etc/nginx/ssl/fqdn/fullchain.pem;
        ssl_certificate_key   /etc/nginx/ssl/fqdn/privkey.pem;
        ssl_session_timeout 1d;
        ssl_session_cache shared:SSL:50m;
        ssl_session_tickets off;
        ssl_protocols TLSv1.2;
        ssl_ciphers 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256';
        ssl_prefer_server_ciphers on;
        add_header Strict-Transport-Security max-age=15768000;
        ssl_stapling on;
        ssl_stapling_verify on;
        resolver 130.60.168.4;

  chunked_transfer_encoding on;

  location / {
    proxy_pass http://127.0.0.1:3000/;
    proxy_set_header Host $http_host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

    proxy_set_header X-Forwarded-Proto $scheme;

    proxy_buffering off;
    proxy_request_buffering off;
  }

  location /api/ws {
    proxy_pass http://127.0.0.1:3000/api/ws;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header Origin "";
  }
}
```

**Comment by tapasmallickin** [comment]
Thanks JtheBAB for your response.
There are some several fundamental differences between your working setup and our.

1. semaphore is running as docker container exposed on tcp/4000 (localhost:4000) instead of tcp/3000
2. semaphore is configured with "SEMAPHORE_WEB_ROOT: /semaphore"

p.s. It works fine if we hit http://local_URL:4000 

Here is the snippet from nginx.conf:

# Ansible Semaphore Application Server
location /semaphore {
        proxy_pass http://127.0.0.1:4000/;
        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_buffering off;
        proxy_request_buffering off;
        #rewrite ^/semaphore/(.*) /$1 break;
        #proxy_redirect off;
}
location /semaphore/api/ws {
        proxy_pass http://127.0.0.1:4000/api/ws;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Origin "";
        #rewrite ^/semaphore/(.*) /$1 break;
        #proxy_redirect off;
}

Any suggestion?

**Reply by JtheBAB** [reply]
I tried it also first als subpath. I couldn’t get it run :/

**Reply by tapasmallickin** [reply]
Basically, we have few more webservices behind same nginx reverse proxy, hence location/subpath is a practical approach in our environment :(

**Reply by tapasmallickin** [reply]
May I expect any help here?
